File: challenging/currency-exchange.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"time"
)

type Output struct {
	Date           string  `json:"date"`
	BaseCurrency   string  `json:"base_currency"`
	TargetCurrency string  `json:"target_currency"`
	ExchangeRate   float64 `json:"exchange_rate"`
}

func main() {
	var input, output string
	if len(os.Args) < 3 {
		input = "input.json"
		output = "output.json"
	} else {
		input = os.Args[1]
		output = os.Args[2]
	}

	clean(input, output)
	curate(input, output)
}

func clean(input, output string) {
	inputFile, err := os.Open(input)
	if err != nil {
		fmt.Println(err)
		return
	}
	defer inputFile.Close()

	outputFile, err := os.Create(output)
	if err != nil {
		fmt.Println(err)
		return
	}
	defer outputFile.Close()

	outputData := Output{"", "EUR", "", 0.0}
	decoder := json.NewDecoder(inputFile)
	encoder := json.NewEncoder(outputFile)

	for decoder.More() {
		var input map[string]map[string]float64
		err := decoder.Decode(&input)
		if err != nil {
			fmt.Println(err)
			return
		}

		for date, currencies := range input {
			outputData.Date = date
			for currency, exchangeRate := range currencies {
				outputData.TargetCurrency = currency
				outputData.ExchangeRate = exchangeRate
				err := encoder.Encode(outputData)
				if err != nil {
					fmt.Println(err)
					return
				}
			}
		}
	}
}

func curate(input, output string) {
	inputFile, err := os.Open(input)
	if err != nil {
		fmt.Println(err)
		return
	}
	defer inputFile.Close()

	outputFile, err := os.Create(output)
	if err != nil {
		fmt.Println(err)
		return
	}
	defer outputFile.Close()

	outputData := make(map[string][]Output)
	decoder := json.NewDecoder(inputFile)
	encoder := json.NewEncoder(outputFile)

	for decoder.More() {
		var input []Output
		err := decoder.Decode(&input)
		if err != nil {
			fmt.Println(err)
			return
		}

		for _, data := range input {
			if _, ok := outputData[data.TargetCurrency]; !ok {
				outputData[data.TargetCurrency] = []Output{}
			}
			outputData[data.TargetCurrency] = append(outputData[data.TargetCurrency], data)
		}
	}

	for _, data := range outputData {
		sort.Slice(data, func(i, j int) bool {
			return data[i].Date < data[j].Date
		})

		lastExchangeRate := data[0].ExchangeRate
		lastDate, _ := time.Parse("2006-01-02", data[0].Date)

		for index, _ := range data {
			if !nextDateExist(lastDate, data) {
				output := Output{"", "EUR", "", 0.0}

				output.Date = lastDate.AddDate(0, 0, 1).Format("2006-01-02")
				output.TargetCurrency = data[index].TargetCurrency
				output.ExchangeRate = lastExchangeRate

				data = append(data, output)
			}

			lastDate, _ = time.Parse("2006-01-02", data[index].Date)
			lastExchangeRate = data[index].ExchangeRate
		}

		for _, data := range data {
			err := encoder.Encode(data)
			if err != nil {
				fmt.Println(err)
				return
			}
		}
	}
}

func nextDateExist(date time.Time, data []Output) bool {
	nextDate := date.AddDate(0, 0, 1).Format("2006-01-02")
	for _, d := range data {
		if d.Date == nextDate {
			return true
		}
	}
	return false
}

File: challenging/currency-exchange.py
import json
import datetime


def clean(input_file, output_file):
    output = {"date": "", "base_currency": "EUR",
              "target_currency": "", "exchange_rate": 0.0}

    with open(input_file) as f:
        currency_data = json.load(f)

    with open(output_file, 'w') as f:
        for date in currency_data:
            for key in currency_data[date].keys():
                output['date'] = date
                output['target_currency'] = key
                output['exchange_rate'] = currency_data[date][key]
                json.dump(output, f)
                f.write('\n')


def curate(input_file, output_file):
    currency_data = {}

    with open(input_file) as f:
        raw_input_data = list(f)

    for json_str in raw_input_data:
        data = json.loads(json_str)

        if data['target_currency'] not in currency_data.keys():
            currency_data[data['target_currency']] = []
        currency_data[data['target_currency']].append(data)

    for currency in currency_data.keys():
        currency_data[currency].sort(
            key=lambda json: datetime.datetime.fromisoformat(json['date']))
        last_exchange_rate = currency_data[currency][0]['exchange_rate']
        last_date = currency_data[currency][0]['date']

        for index in range(1, len(currency_data[currency])):
            data = currency_data[currency][index]

            if next_date_exist(last_date, currency_data[currency]):
                last_exchange_rate = data['exchange_rate']
                last_date = data['date']
            else:
                output = {"date": "", "base_currency": "EUR",
                          "target_currency": "", "exchange_rate": 0.0}
                output['date'] = (datetime.datetime.fromisoformat(
                    last_date) + datetime.timedelta(days=1)).strftime('%Y-%m-%d')
                output['target_currency'] = currency
                output['exchange_rate'] = last_exchange_rate

                currency_data[currency].append(output)

                last_date = output['date']

    with open(output_file, 'w') as f:
        for currency in currency_data:
            for data in currency_data[currency]:
                json.dump(data, f)
                f.write('\n')

def next_date_exist(date, data):
    next_date = (datetime.datetime.fromisoformat(date) +
                 datetime.timedelta(days=1)).strftime('%Y-%m-%d')
    if next((item for item in data if item["date"] == next_date), False):
        return True
    else:
        return False
